sdcm/provision/gce/virtual_machine_provider.py

```diff
@@ -0,0 +1,140 @@
+        found_instance = self._get_node_from_cache(node_name=instance_definition.name)
+
+        if found_instance:
+            LOGGER.info("Requested the creation of instance with name %s but a "
+                        "instance with that name already exists. Returning the "
+                        "already existing instance.", instance_definition.name)
+            return found_instance
+
+        instance_params = {
+            "name": instance_definition.name,
+            "size": instance_definition.type,
+            "image": instance_definition.image_id,
+            "ex_network": params.get("gce_network"),
+            "ex_disks_gce_struct": self._disk_struct_provider.get_disks_struct(disk_struct_args, params),
+            "ex_metadata": self._metadata_provider.get_ex_metadata(metadata_args, params),
+            "ex_service_accounts": self._service_accounts,
+            "ex_preemptible": pricing_model == pricing_model.SPOT
+        }
+
+        LOGGER.info("Going to create the instance using the following instance params:\n%s", instance_params)
+
+        try:
+            new_node: Node = self._gce_service.create_node(**instance_params)
+            self._vm_node_cache.append(new_node)
+            return new_node
+        except Exception as exc:
+            raise VMProvisionError("Could not provision VM. Root error:\n%s") from exc
+
+    def _get_node_from_cache(self, node_name: str) -> Node | None:
+        found_node = [node for node in self._vm_node_cache if node.name == node_name]
+
+        if found_node:
+            return found_node[0]
+        return None
+
+    def _get_random_location_in_region(self) -> str:
+        return random.choice(self._get_available_locations())
+
+    def _get_available_locations(self) -> list[str]:
+        locations = [item.name for item in self._gce_service.list_locations() if item.name.startswith(self._region)]
+        return locations
+
+    def add_tags(self, vm_instance_name: str, tags: dict):
+        instance_to_tag = [node for node in self.list_running_instances() if node.name == vm_instance_name]
+        if not instance_to_tag:
+            raise VMProvisionError(f"Can't find instance to add tags too. Instance name requested: {vm_instance_name}")
+
+        LOGGER.info("Sanitized tags: %s", self._sanitize_tags(tags))
+        self._gce_service.ex_set_node_labels(instance_to_tag[0], self._sanitize_tags(tags))
+        LOGGER.info("Tags as node labels added to node: %s", instance_to_tag[0].name)
+
+    def list_instances(self) -> list[Node]:
+        return self._gce_service.list_nodes()
+
+    def list_running_instances(self) -> list[Node]:
+        running_instances = [instance for instance in self.list_instances() if instance.state in NodeState.RUNNING]
+        return running_instances
+
+    def destroy_instance(self, instance_name: str) -> bool:
+        instance_to_destroy = self._get_node_from_cache(instance_name)
+
+        if instance_to_destroy:
```

Comment:
In Azure we log Warning if user tries destroy instance that does not exist:
`LOGGER.warning("Instance %s does not exist. Shouldn't have called it", name)`

Reply:
Fixed (warning log added)